### PR TITLE
Implement TMC2209 and stepper generation

### PR DESCRIPTION
**Pull request by LoaDy588**

### CODE REVIEW

File: firmware/.vscode/c_cpp_properties.json
Comment (LoaDy588) [inline firmware/.vscode/c_cpp_properties.json:1]: Remove this file - unneeded vscode config file

File: cube_lib/include/hardware.hpp [hunk 1]
```diff
@@ -7,15 +7,18 @@
...
     endstop_triggered,
-    i2c_transfer_error,
+    i2c_transfer_error, ///< timeout or internal I2C error
+    uart_transmit_error,
     spi_transfer_error,
     gpio_set_error,
     gpio_read_error,
-    gpio_wrong_mode
+    gpio_wrong_mode,
+    param_nonexistant,
```
Comment (LoaDy588) [inline cube_lib/include/hardware.hpp]: would change this to `param_id_invalid`

File: cube_lib/source/planner.cpp [hunk 2]
```diff
@@ -98,8 +98,8 @@ planner_result planner::do_move(const point& pos) {
 
     case planner_machine::corexy:
         //the 0.5f should not be here? we'll see
```
Comment (LoaDy588) [inline cube_lib/source/planner.cpp]: this comment is now even more confusing

File: firmware/include/core_xy.hpp
Comment (LoaDy588) [inline firmware/include/core_xy.hpp:1]: To consider: create abstract `stepper_generation` base class to define interfaces, for easier implementation of future machine types

File: firmware/source/core_xy.cpp [hunk 3]
```diff
@@ -0,0 +1,90 @@
+#include <numbers>
+#include "peripherals.hpp"
+#include "main.hpp"
+#include "config.hpp"
+#include "core_xy.hpp"
+
+
+namespace cube_hw {
+
+constexpr float PI = std::numbers::pi_v<float>;
```
Comment (LoaDy588) [inline firmware/source/core_xy.cpp:7]: if these constants are just local, I would move them to anonymous namespace
another option: create `firmware/include/constants.hpp` for all various constants

File: firmware/include/stepper_generator.hpp [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    const MotorPins& pins;
+    int dma_burst_padding;
+    uint32_t dma_burst_length;
+    const unsigned _steps_per_mm;
+    motor_state _state = motor_state::RESET;
+    float ratio = 1.0f;
+    motor_config config;
+
+    // Dma
+    std::vector<uint16_t> acceleration_mem;
```
Comment (LoaDy588) [inline firmware/include/stepper_generator.hpp:47]: I would strongly prefer static allocation for this. Or atleast pre-allocate in constructor and then never resize

File: firmware/.vscode/c_cpp_properties.json
Comment (Gaglii) [reply firmware/.vscode/c_cpp_properties.json:1]: removed

File: cube_lib/source/planner.cpp (hunk 2 above)
Comment (Gaglii) [reply cube_lib/source/planner.cpp]: comment removed

File: cube_lib/include/hardware.hpp (hunk 1 above)
Comment (Gaglii) [reply cube_lib/include/hardware.hpp]: renamed

File: firmware/source/core_xy.cpp (hunk 3 above)
Comment (Gaglii) [reply firmware/source/core_xy.cpp:7]: moved to anonymous namespace

File: firmware/include/stepper_generator.hpp (hunk 4 above)
Comment (Gaglii) [reply firmware/include/stepper_generator.hpp:47]: The vector is now resized only once

**Pull request merged**